README.md
jaymap
======

JMAP client built in asyncio

[![version](https://img.shields.io/pypi/v/jaymap.svg)](https://pypi.org/project/jaymap)
[![license](https://img.shields.io/pypi/l/jaymap.svg)](https://github.com/jreese/jaymap/blob/master/LICENSE)


License
-------

jaymap is copyright [John Reese](https://jreese.sh), and licensed under
the MIT license.  I am providing code in this repository to you under an open
source license.  This is my personal repository; the license you receive to
my code is from me and not from my employer. See the `LICENSE` file for details.


CHANGELOG.md
jaymap
======

v0.1.0a1
--------

Alpha release

- Basic support for custom datatypes matching JMAP Core spec
- Support for serializing and deserializing datatypes to dict/json
- Simple client that can fetch session/status and make arbitrary requests
- Smoke test that takes domain/user/pass and fetches a single email

```
$ git shortlog -s baad05bf601d392f00b8b637fc6be1e9adbdcc40...v0.1.0a1
     1	John Reese
```



jaymap/tests/smoke.py
# Copyright 2021 John Reese
# Licensed under the MIT license

import asyncio
import logging
import sys

import click

from jaymap.client import JMAP
from jaymap.types.core import Request, Invocation, Capabilities
from jaymap.types.mail import Email


@click.command()
@click.option("--domain", prompt=True)
@click.option("--username", prompt=True)
@click.option("--password", prompt=True, hide_input=True)
def main(domain: str, username: str, password: str):
    logging.basicConfig(level=logging.DEBUG, stream=sys.stdout)

    async def inner():
        async with JMAP(domain, username, password) as client:
            click.echo(client.session)
            account_id = client.session.primary_accounts[Capabilities.MAIL]
            methods = [
                (
                    "Mailbox/query",
                    {"accountId": account_id, "filter": {"role": "inbox"}},
                    "c1",
                ),
                (
                    "Mailbox/get",
                    {
                        "accountId": account_id,
                        "#ids": {
                            "resultOf": "c1",
                            "name": "Mailbox/query",
                            "path": "/ids",
                        },
                    },
                    "c2",
                ),
                (
                    "Email/query",
                    {
                        "accountId": account_id,
                        "filter": {
                            "inMailbox": "d00b1aa9-ce5c-4775-ba79-e8a8d1a042fd",
                        },
                        "limit": 2,
                    },
                    "c3",
                ),
                (
                    "Email/get",
                    {
                        "accountId": account_id,
                        "#ids": {
                            "resultOf": "c3",
                            "name": "Email/query",
                            "path": "/ids",
                        },
                    },
                    "c3",
                ),
            ]
            req = Request(
                using=client.capabilities,
                method_calls=methods,
            )
            res = await client.request(req)
            emails = res.method_responses[-1][1]["list"]
            emails = Email.from_list(emails)
            click.secho(f"{emails!r}")

    asyncio.run(inner())


if __name__ == "__main__":
    main()  # pylint: disable=all


jaymap/tests/types.py
# Copyright 2021 John Reese
# Licensed under the MIT license

from unittest import TestCase

from jaymap.types import base, core

TEST_CAPABILITIES = {core.Capabilities.CORE: {"maxConcurrentRequests": 8}}
TEST_OBJECTS = (
    core.Account(
        name="foo@bar.com",
        is_personal=True,
        is_read_only=False,
        account_capabilities=TEST_CAPABILITIES,
    ),
    core.Session(
        capabilities=TEST_CAPABILITIES,
        accounts={
            core.Id("f123"): core.Account(
                name="foo@bar.com",
                is_personal=True,
                is_read_only=False,
                account_capabilities=TEST_CAPABILITIES,
            ),
        },
        primary_accounts={core.Capabilities.CORE: core.Id("f123")},
        username="foo@bar.com",
        api_url="https://localhost:4433/api/",
        download_url="https://localhost:4433/download/",
        upload_url="https://localhost:4433/upload/",
        event_source_url="https://localhost:4433/events/",
        state="aoeu1234",
    ),
    core.Request(
        using=[core.Capabilities.CORE],
        method_calls=[
            core.Invocation("func1", {"name": "foo"}, "c1"),
            core.Invocation("func2", {"name": "foo"}, "c2"),
        ],
    ),
)


class CoreTypesTest(TestCase):
    def test_id_class(self):
        for value in ("a", "1", "-", "u1234", "23a890Z_23-"):
            with self.subTest(f"valid id {value}"):
                id = core.Id(value)
                self.assertIsInstance(id, core.Id)
                self.assertEqual(id, value)

        for value in ("", "foo*bar", "foo()", "$100"):
            with self.subTest(f"invalid id {value}"):
                with self.assertRaisesRegex(ValueError, r"invalid Id"):
                    core.Id(value)

    def test_int_class(self):
        for value in (
            0,
            1,
            -1,
            1000,
            -10000,
            2 ** 53 - 1,
            -(2 ** 53) + 1,
            "0",
            "-1000",
        ):
            with self.subTest(f"valid int {value}"):
                id = core.Int(value)
                self.assertIsInstance(id, core.Int)
                self.assertEqual(id, int(value))

        for value in (2 ** 53, -(2 ** 53), 2 ** 54, -(2 ** 54)):
            with self.subTest(f"invalid int {value}"):
                with self.assertRaisesRegex(ValueError, r"Int -?\d+ [<>]"):
                    core.Int(value)

    def test_unsigned_int_class(self):
        for value in (0, 1, 1000, 1000000, 2 ** 53 - 1, "0", "1000"):
            with self.subTest(f"valid unsigned int {value}"):
                id = core.UnsignedInt(value)
                self.assertIsInstance(id, core.UnsignedInt)
                self.assertEqual(id, int(value))

        for value in (-1, -10, -10000000, -(2 ** 53), "-1", "-1000"):
            with self.subTest(f"invalid unsigned int {value}"):
                with self.assertRaisesRegex(ValueError, r"UnsignedInt -?\d+ [<>]"):
                    core.UnsignedInt(value)

    def test_is_datatype(self):
        for obj in TEST_OBJECTS:
            with self.subTest(obj.__class__.__name__):
                self.assertTrue(base.is_datatype(obj))
                self.assertTrue(base.is_datatype(obj.__class__))

    def test_datatype_encode_decode(self):
        for obj in TEST_OBJECTS:
            with self.subTest(obj.__class__.__name__):
                text = obj.to_json(indent=4)
                obj2 = obj.__class__.from_json(text)
                self.assertEqual(obj.__class__, obj2.__class__)
                self.assertEqual(obj, obj2)


jaymap/tests/__init__.py
# Copyright 2021 John Reese
# Licensed under the MIT license

from .types.base import BaseTypes
from .types.core import CoreTypes


jaymap/client.py
# Copyright 2021 John Reese
# Licensed under the MIT License

import asyncio
import json
import logging
from contextlib import AsyncExitStack
from dataclasses import dataclass, field
from typing import Dict, List

import aiohttp

from jaymap.types.core import Session, Request, Response

LOG = logging.getLogger(__name__)


def log_json(message: str, text: str) -> None:
    data = json.loads(text)
    LOG.debug("%s\n%s", message, json.dumps(data, indent=4, sort_keys=True))


@dataclass(eq=False, order=False)
class JMAP:
    domain: str  # jmap.fastmail.com
    username: str  # user@fastmail.com
    password: str = field(repr=False)  # hunter42
    session: Session = field(init=False)

    capabilities: List[str] = field(default_factory=list)
    account_id: str = ""

    stack: AsyncExitStack = field(init=False, repr=False)
    client: aiohttp.ClientSession = field(init=False, repr=False)

    @property
    def client_headers(self) -> Dict[str, str]:
        return {}

    def __init__(self, domain: str, username: str, password: str):
        self.domain = domain
        self.username = username
        self.password = password

    async def __aenter__(self) -> "JMAP":
        self.stack = AsyncExitStack()
        basic_auth = aiohttp.BasicAuth(self.username, self.password)
        self.client = await self.stack.enter_async_context(
            aiohttp.ClientSession(auth=basic_auth)
        )
        await self.update_session()
        return self

    async def __aexit__(self, *args) -> None:
        await self.stack.aclose()

    async def update_session(self) -> Session:
        discovery_url = f"https://{self.domain}/.well-known/jmap"
        LOG.debug("sending auto-discovery request to %s", discovery_url)
        async with self.client.get(discovery_url) as response:
            text = await response.text("utf-8")
            if response.status != 200:
                raise ValueError(f"response error {response.status}: {text}")

            log_json("auto-discovery response", text)
            self.session = Session.from_json(text)
            self.capabilities = list(self.session.capabilities)

        return self.session

    async def request(self, req: Request) -> Response:
        api_url = self.session.api_url
        text = req.to_json()
        LOG.debug("JMAP.request(%s)\n  %s\n  %s", repr(req), api_url, text)
        async with self.client.post(api_url, data=text) as response:
            text = await response.text("utf-8")
            if response.status != 200:
                raise ValueError(f"response error {response.status}: {text}")

            log_json("API response", text)
            data = Response.from_json(text)

            if data.session_state != self.session.state:
                LOG.debug("session state changed, triggering auto-discovery")
                asyncio.create_task(self.update_session())

            return data


jaymap/types/base.py
# Copyright 2021 John Reese
# Licensed under the MIT License

"""
Base datatype implementation with encoding/decoding support
"""

import json
import re
import sys
from dataclasses import dataclass, field, is_dataclass, asdict, fields
from textwrap import indent
from typing import (
    get_type_hints,
    Union,
    Dict,
    Any,
    Mapping,
    List,
    Tuple,
    Optional,
    Sequence,
    Set,
    Tuple,
    Type,
    TypeVar,
    cast,
)

from stringcase import camelcase
from typing_inspect import (
    is_generic_type,
    is_optional_type,
    is_union_type,
    get_origin,
    get_args,
    get_generic_bases,
)

T = TypeVar("T")

INDENT = "    "
PRIMITIVES = (bool, float, int, bytes, str, type(None), Any)


def is_primitive(ftype: Type) -> bool:
    try:
        return ftype in PRIMITIVES or issubclass(ftype, PRIMITIVES)
    except TypeError:
        return False


def decode(obj: Any, hint: Type) -> Any:
    ftype = hint
    if ftype in PRIMITIVES:
        return obj

    if is_optional_type(ftype):
        if obj is None:
            return None

        targs = [t for t in get_args(ftype) if t is not type(None)]
        if len(targs) > 1:
            raise NotImplementedError(f"can't decode union types ({targs})")
        ftype = targs[0]
    elif is_union_type(ftype):
        raise NotImplementedError(f"can't decode union types ({get_args(ftype)})")

    if is_primitive(ftype):
        return ftype(obj)

    if is_datatype(ftype):
        if not isinstance(obj, dict):
            raise TypeError(f"invalid data {obj!r} for {ftype}")

        kwargs: Dict[str, Any] = {}

        namespace = sys.modules[ftype.__module__].__dict__
        for fname, ft in get_type_hints(ftype, namespace).items():
            key = camelcase(fname.strip("_"))

            if key in obj:
                kwargs[fname] = decode(obj[key], ft)

        return ftype(**kwargs)

    if is_generic_type(ftype):
        origin = get_origin(ftype)
        bases = get_generic_bases(ftype)
        targs = get_args(ftype)

        while origin is None and bases:
            if len(bases) > 1:  # pragma: nocover
                raise NotImplementedError(f"can't decode multiple bases {ftype}")
            ftype = bases[0]
            origin = get_origin(ftype)
            bases = get_generic_bases(ftype)
            targs = get_args(ftype)

        if origin in (dict, Dict, Mapping):
            if not is_primitive(targs[0]):
                raise NotImplementedError(f"can't decode object keys {ftype}")
            ftype = targs[1]

            if ftype in PRIMITIVES:
                return dict(obj)
            else:
                return {k: decode(v, ftype) for k, v in obj.items()}

        if origin in (tuple, Tuple):
            return tuple(decode(v, ftype) for v, ftype in zip(obj, targs))

        if origin in (set, Set):
            ftype = targs[0]

            if ftype in PRIMITIVES:
                return set(obj)
            else:
                return set(decode(v, ftype) for v in obj)

        if origin in (list, List):
            ftype = targs[0]

            if ftype in PRIMITIVES:
                return list(obj)
            else:
                return [decode(v, ftype) for v in obj]

    raise NotImplementedError(f"failed to decode {obj} as type {hint}")


def encode(obj: Any, hint: Type) -> Any:
    ftype = hint
    if obj is None or is_primitive(ftype):
        return obj

    if is_optional_type(ftype):
        targs = [t for t in get_args(ftype) if t is not type(None)]
        if len(targs) > 1:
            raise NotImplementedError(f"can't encode union types ({targs})")
        ftype = targs[0]
    elif is_union_type(ftype):
        raise NotImplementedError(f"can't encode union types ({get_args(ftype)})")

    if is_generic_type(ftype):
        origin = get_origin(ftype)
        bases = get_generic_bases(ftype)
        targs = get_args(ftype)

        while origin is None and bases:
            if len(bases) > 1:  # pragma: nocover
                raise NotImplementedError(f"can't encode multiple bases {ftype}")
            ftype = bases[0]
            origin = get_origin(ftype)
            bases = get_generic_bases(ftype)
            targs = get_args(ftype)

        if origin in (dict, Dict, Mapping):
            if not is_primitive(targs[0]):
                raise NotImplementedError(f"can't encode object keys {ftype}")
            ftype = targs[1]

            if is_primitive(ftype):
                return obj
            else:
                return {
                    camelcase(k.strip("_")): encode(v, ftype) for k, v in obj.items()
                }

        if origin in (tuple, Tuple):
            return [encode(v, ftype) for v, ftype in zip(obj, targs)]

        if origin in (set, list, Sequence, Set):
            ftype = targs[0]

            if is_primitive(ftype):
                return list(obj)
            else:
                return [encode(v, ftype) for v in obj]

    if is_primitive(ftype):
        return obj

    if isinstance(obj, Datatype):
        result: Dict[str, Any] = {}

        namespace = sys.modules[ftype.__module__].__dict__
        for fname, ftype in get_type_hints(obj, namespace).items():
            key = camelcase(fname.strip("_"))
            value = getattr(obj, fname)

            result[key] = encode(value, ftype)

        return result

    raise NotImplementedError(f"failed to encode {hint}({type(obj)!r})")


class Datatype:
    """Dataclass derivative with json encode/decode capabilities"""

    def __init_subclass__(cls):
        dataclass(repr=False)(cls)

    def __repr__(self) -> str:
        parts: List[str] = []
        for field in fields(self):
            value = getattr(self, field.name)
            if isinstance(value, list) and value:
                inner = "\n".join(f"{v!r}," for v in value)
                part = f"[\n{indent(inner, INDENT)}\n]"
            elif isinstance(value, dict) and value:
                inner = "\n".join(f"{k!r}: {v!r}," for k, v in value.items())
                part = f"{{\n{indent(inner, INDENT)}\n}}"
            else:
                part = repr(value)
            parts.append(f"{field.name}={part},")
        inner = indent("\n".join(parts), INDENT)
        return f"{self.__class__.__name__}(\n{inner}\n)"

    @classmethod
    def from_dict(cls: Type[T], data: Dict[str, Any]) -> T:
        return cast(cls, decode(data, cls))

    @classmethod
    def from_list(cls: Type[T], data: List[Dict[str, Any]]) -> List[T]:
        return [cls.from_dict(v) for v in data]

    @classmethod
    def from_json(cls: Type[T], text: str) -> T:
        return cls.from_dict(json.loads(text))

    def to_dict(self) -> Dict[str, Any]:
        return encode(self, self.__class__)

    def to_json(self, **kwargs: Any) -> str:
        kwargs.setdefault("sort_keys", True)
        return json.dumps(self.to_dict(), **kwargs)


def is_datatype(obj: Any) -> bool:
    return is_dataclass(obj) and (
        isinstance(obj, Datatype) or (type(obj) == type and issubclass(obj, Datatype))
    )


jaymap/types/core.py
# Copyright 2021 John Reese
# Licensed under the MIT License

# pylint: disable=too-few-public-methods

"""
https://jmap.io/spec-core.html
"""

import re
import time
from datetime import datetime, timezone
from typing import Union, Dict, Tuple, Any, List, Optional, Type

from jaymap.types.base import Datatype, T


class Capabilities:
    CORE = "urn:ietf:params:jmap:core"
    MAIL = "urn:ietf:params:jmap:mail"
    CONTACTS = "urn:ietf:params:jmap:contacts"
    CALENDAR = "urn:ietf:params:jmap:calendars"


class Id(str):
    _VALID = re.compile(r"[A-Za-z0-9_-]{1,255}")

    def __new__(cls, value: str):
        if not cls._VALID.fullmatch(value):
            raise ValueError(f"invalid Id {value!r}")
        return str.__new__(cls, value)  # type: ignore

    def __repr__(self) -> str:
        return f"Id({super().__repr__()})"


class Int(int):
    _MIN = -(2 ** 53) + 1
    _MAX = 2 ** 53 - 1

    def __new__(cls, value: Union[str, int]):
        if isinstance(value, str):
            value = int(value)
        if value < cls._MIN:
            raise ValueError(f"{cls.__name__} {value!r} < {cls._MIN}")
        if value > cls._MAX:
            raise ValueError(f"{cls.__name__} {value!r} > {cls._MAX}")
        return int.__new__(cls, value)  # type: ignore

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({super().__repr__()})"


class UnsignedInt(Int):
    _MIN = 0


class Date(str):
    """"2014-10-30T06:12:00+05:00"""

    DT_PARSE = r"%Y-%m-%dT%H:%M:%S%z"
    DT_FMT = r"%Y-%m-%dT%H:%M:%S%z"
    UTC_FMT = r"%Y-%m-%dT%H:%M:%SZ"

    def __new__(cls, value: str, dt: Optional[datetime] = None):
        return super().__new__(cls, value)

    def __init__(self, value: str, dt: Optional[datetime] = None):
        if dt is None:
            self.datetime = datetime.strptime(value, self.DT_PARSE)
        else:
            self.datetime = dt

    @classmethod
    def from_datetime(cls: Type[T], source: datetime) -> "T":
        if source.tzinfo in (None, timezone.utc):
            value = source.strftime(cls.UTC_FMT)
        else:
            value = source.strftime(cls.DT_FMT)
            value = f"{value[:-2]}:{value[-2:]}"
        return cls(value, source)


class UTCDate(Date):
    """"2014-10-30T06:12:00Z"""

    DTF = r"%Y-%m-%dT%H:%M:%SZ"
    DTP = r"%Y-%m-%dT%H:%M:%S%z"

    def __init__(self, value: str, dt: Optional[datetime] = None):
        if dt and dt.tzinfo not in (None, timezone.utc):
            raise ValueError(f"invalid timezone {dt.tzinfo} for UTCDate")
        super().__init__(value, dt)
        if self.datetime.tzinfo not in (None, timezone.utc):
            raise ValueError(
                f"invalid parsed timezone {self.datetime.tzinfo} for UTCDate"
            )

    @classmethod
    def from_datetime(cls: Type[T], source: datetime) -> "T":
        if source.tzinfo not in (None, timezone.utc):
            raise ValueError(f"invalid timezone {source.tzinfo} for UTCDate")
        return super().from_datetime(source)


class Account(Datatype):
    name: str
    is_personal: bool
    is_read_only: bool
    account_capabilities: Dict[str, Any]


class Session(Datatype):
    capabilities: Dict[str, Any]
    accounts: Dict[Id, Account]
    primary_accounts: Dict[str, Id]
    username: str
    api_url: str
    download_url: str
    upload_url: str
    event_source_url: str
    state: str


class Invocation(Tuple[str, Dict[str, Any], str]):
    def __new__(cls, name: str, arguments: Dict[str, Any], call_id: str):
        return tuple.__new__(cls, (name, arguments, call_id))


class Request(Datatype):
    using: List[str]
    method_calls: List[Invocation]
    created_ids: Optional[Dict[Id, Id]] = None


class Response(Datatype):
    method_responses: List[Invocation]
    session_state: str
    created_ids: Optional[Dict[Id, Id]] = None


jaymap/types/mail.py
# Copyright 2021 John Reese
# Licensed under the MIT License

"""
https://jmap.io/spec-mail.html
"""

from typing import Optional, Any, List, Dict

from jaymap.types.base import Datatype, field
from jaymap.types.core import Id, UnsignedInt, UTCDate


class MailboxRights(Datatype):
    may_read_items: bool
    may_add_items: bool
    may_remove_items: bool
    may_set_seen: bool
    may_set_keywords: bool
    may_create_child: bool
    may_rename: bool
    may_delete: bool
    may_submit: bool


class Mailbox(Datatype):
    id: Id
    name: str
    parent_id: Optional[Id]
    role: Optional[str]
    sort_order: UnsignedInt
    total_emails: UnsignedInt
    unread_emails: UnsignedInt
    total_threads: UnsignedInt
    unread_threads: UnsignedInt
    my_rights: MailboxRights
    is_subscribed: bool


class Thread(Datatype):
    id: Id
    email_ids: List[Id]


class EmailAddress(Datatype):
    name: Optional[str]
    email: str


class EmailAddressGroup(Datatype):
    name: Optional[str]
    addresses: List[EmailAddress]


class EmailHeader(Datatype):
    name: str
    value: str


class EmailBodyPart(Datatype):
    part_id: Optional[str]
    blob_id: Optional[Id]
    size: UnsignedInt
    name: Optional[str]
    type: str
    charset: Optional[str]
    disposition: Optional[str]
    cid: Optional[str]
    language: Optional[List[str]]
    location: Optional[str]
    headers: List[EmailHeader] = field(default_factory=list)
    sub_parts: Optional[List["EmailBodyPart"]] = field(default_factory=list)


class EmailBodyValue(Datatype):
    value: str
    is_encoding_problem: bool = False
    is_truncated: bool = False


class Email(Datatype):
    id: Id
    blob_id: Id
    thread_id: Id
    mailbox_ids: Dict[Id, bool]
    keywords: Dict[str, bool]
    size: UnsignedInt
    received_at: UTCDate
    message_id: Optional[List[str]]
    in_reply_to: Optional[List[str]]
    references: Optional[List[str]]
    sender: Optional[List[EmailAddress]]
    from_: Optional[List[EmailAddress]]
    to: Optional[List[EmailAddress]]
    cc: Optional[List[EmailAddress]]
    bcc: Optional[List[EmailAddress]]
    reply_to: Optional[List[EmailAddress]]
    subject: Optional[str]
    sent_at: Optional[str]
    body_values: Dict[str, EmailBodyValue]
    text_body: List[EmailBodyPart]
    html_body: List[EmailBodyPart]
    attachments: List[EmailBodyPart]
    has_attachment: bool
    preview: str
    body_structure: Optional[EmailBodyPart] = None
    headers: List[EmailHeader] = field(default_factory=list)


class SearchSnippet(Datatype):
    email_id: Id
    subject: Optional[str]
    preview: Optional[str]


class Identity(Datatype):
    id: Id
    name: str
    email: str
    reply_to: Optional[List[str]]
    bcc: Optional[List[str]]
    text_signature: str
    html_signature: str
    may_delete: bool


class Address(Datatype):
    email: str
    parameters: Any


class Envelope(Datatype):
    mail_from: str
    rcpt_to: List[Address]


class DeliveryStatus(Datatype):
    smtp_reply: str
    delivered: str
    displayed: str


class EmailSubmission(Datatype):
    id: Id
    identity_id: Id
    email_id: Id
    thread_id: Id
    envelope: Optional[Envelope]
    send_at: UTCDate
    undo_status: str
    delivery_status: Optional[DeliveryStatus]
    dsn_blob_ids: List[Id] = field(default_factory=list)
    mdn_blob_ids: List[Id] = field(default_factory=list)


class VacationResponse(Datatype):
    id: Id
    is_enabled: bool
    from_date: Optional[UTCDate]
    to_date: Optional[UTCDate]
    subject: Optional[str]
    text_body: Optional[str]
    html_body: Optional[str]


jaymap/types.py
# Copyright 2021 John Reese
# Licensed under the MIT License

import re
from dataclasses import dataclass
from typing import NewType, Union, Dict, Any, List, Tuple, Optional, Type

from dataclasses_json import LetterCase, DataClassJsonMixin
from dataclasses_json.api import _process_class


def datatype(cls: Type) -> Type:
    cls = dataclass(cls)
    cls = _process_class(cls, LetterCase.CAMEL, None)
    return cls


class Id(str):
    _VALID = re.compile(r"[A-Za-z0-9_-]{1,255}")

    def __new__(cls, value: str):
        if not cls._VALID.fullmatch(value):
            raise ValueError(f"invalid Id {value!r}")
        return str.__new__(cls, value)  # type: ignore


class Int(int):
    _MIN = -(2 ** 53) + 1
    _MAX = 2 ** 53 - 1

    def __new__(cls, value: Union[str, int]):
        if isinstance(value, str):
            value = int(value)
        if value < cls._MIN:
            raise ValueError(f"{cls.__name__} {value!r} < {cls._MIN}")
        if value > cls._MAX:
            raise ValueError(f"{cls.__name__} {value!r} > {cls._MAX}")
        return int.__new__(cls, value)  # type: ignore


class UnsignedInt(Int):
    _MIN = 0


@datatype
class Account(DataClassJsonMixin):
    name: str
    is_personal: bool
    is_read_only: bool
    account_capabilities: Dict[str, Any]


@datatype
class Session(DataClassJsonMixin):
    capabilities: Dict[str, Any]
    accounts: Dict[Id, Account]
    primary_accounts: Dict[str, Id]
    username: str
    api_url: str
    download_url: str
    upload_url: str
    event_source_url: str
    state: str


Invocation = NewType("Invocation", Tuple[str, Dict[str, Any], str])


@datatype
class Request(DataClassJsonMixin):
    using: List[str]
    method_calls: List[Invocation]
    created_ids: Optional[Dict[Id, Id]] = None


@datatype
class Response(DataClassJsonMixin):
    method_responses: List[Invocation]
    session_state: str
    created_ids: Optional[Dict[Id, Id]] = None
